lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -120,42 +122,44 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
           end
         end
 
-        # Update listeners
-        perform_listener_action = proc do |desc, &block|
-          perform_listener_action = proc { |desc, &block| perform_action(desc, &block) }
-        end
+        # Update listeners - THIS IS NOT ATOMIC
         add_listeners = {}
         listeners.each { |l| add_listeners[l[:port]] = l } if listeners
         actual_elb.listeners.each do |listener|
           desired_listener = add_listeners.delete(listener.port)
           if desired_listener
-            if listener.protocol != desired_listener[:protocol]
-              perform_listener_action.call("    update protocol from #{listener.protocol.inspect} to #{desired_listener[:protocol].inspect}'") do
-                listener.protocol = desired_listener[:protocol]
-              end
+
+            # listener.(port|protocol|instance_port|instance_protocol) are immutable for the life
+            # of the listener - must create a new one and delete old one
+            immutable_updates = []
+            if listener.protocol != desired_listener[:protocol].to_sym.downcase
+              immutable_updates << "    update protocol from #{listener.protocol.inspect} to #{desired_listener[:protocol].inspect}"
             end
             if listener.instance_port != desired_listener[:instance_port]
-              perform_listener_action.call("    update instance port from #{listener.instance_port.inspect} to #{desired_listener[:instance_port].inspect}'") do
-                listener.instance_port = desired_listener[:instance_port]
-              end
+              immutable_updates << "    update instance port from #{listener.instance_port.inspect} to #{desired_listener[:instance_port].inspect}"
             end
-            if listener.instance_protocol != desired_listener[:instance_protocol]
-              perform_listener_action.call("    update instance protocol from #{listener.instance_protocol.inspect} to #{desired_listener[:instance_protocol].inspect}'") do
-                listener.instance_protocol = desired_listener[:instance_protocol]
-              end
+            if listener.instance_protocol != desired_listener[:instance_protocol].to_sym.downcase
+              immutable_updates << "    update instance protocol from #{listener.instance_protocol.inspect} to #{desired_listener[:instance_protocol].inspect}"
             end
-            if listener.server_certificate != desired_listener[:server_certificate]
-              perform_listener_action.call("    update server certificate from #{listener.server_certificate} to #{desired_listener[:server_certificate]}'") do
+            if !immutable_updates.empty?
+              perform_action.call(immutable_updates) do
+                listener.delete
```

Comment:
Yup, thats the exact same documentation I was looking at that made me give up trying to figure out how to make it atomic :)